library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/service/GafferPopNamedOperationService.java

```diff
@@ -44,44 +57,61 @@ public Type getType() {
+     */
     @Override
     public CloseableIterator<R> execute(final ServiceCallContext ctx, final Map params) {
         if (params.containsKey("execute")) {
-            final String name = (String) params.get("execute");
-            Iterable<NamedOperationDetail> namedOps = graph.execute(new OperationChain.Builder().first(new GetAllNamedOperations()).build());
-            for (final NamedOperationDetail namedOp : namedOps) {
-                if (namedOp.getOperationName().equals(name)) {
-                    if (namedOp.getOperationChainWithDefaultParams().getOutputClass().equals(Iterable.class)) {
-                        return (CloseableIterator<R>) CloseableIterator.of(graph.execute(new OperationChain.Builder()
-                                .first(new NamedOperation.Builder()
-                                        .name(name)
-                                        .build())
-                                .then(new GenerateObjects.Builder<GafferPopElement>()
-                                        .generator(new GafferPopElementGenerator(graph))
-                                        .build())
-                                .build()).iterator());
-                    } else {
-                        return CloseableIterator.of(Arrays.asList(graph.execute(new OperationChain.Builder()
-                                .first(new NamedOperation.Builder<Void, R>()
-                                        .name(name)
-                                        .build())
-                                .build())).iterator());
-                    }
-                }
-            }
-            throw new IllegalStateException("Named Operation not found");
+            return executeNamedOperation((String) params.get("execute"));
         } else if (params.containsKey("add")) {
-            final Map<String, Object> addParams = (Map) params.get("add");
-            graph.execute(new OperationChain.Builder()
-                    .first(new AddNamedOperation.Builder()
-                            .name((String) addParams.get("name"))
-                            .operationChain((String) addParams.get("opChain"))
-                            .build())
-                    .build());
-            return CloseableIterator.empty();
+            return addNamedOperation((Map<String, String>) params.get("add"));
         } else {
             throw new IllegalStateException("Missing parameter");
         }
     }
 
+    protected CloseableIterator<R> executeNamedOperation(final String name) {
+        // Fetch details for the requested Named Operation or throw an exception if it's not found
+        Iterable<NamedOperationDetail> allNamedOps = graph.execute(new OperationChain.Builder().first(new GetAllNamedOperations()).build());
+        NamedOperationDetail namedOperation = StreamSupport.stream(allNamedOps.spliterator(), false)
+                .filter((NamedOperationDetail nd) -> nd.getOperationName().equals(name)).findFirst()
```

Comment:
Think this is an unnecessary cast to `NamedOperationDetail` as the `Iterable` has already been declared as this type.

Also `.findFirst()` should be on the next line to be consistent with formatting.

Reply:
This isn't a cast. It just provides the full type for the lambda parameter. However, you're correct that this isn't required. I'll change the formatting too.